### ISSUE TITLE
Proposal for an `ngFormChange` event

### ISSUE THREAD

**Issue by CWSpear** [angular/angular#24818]
## I'm submitting a...

<pre><code>[x] Feature request</code></pre>

## Current behavior
Currently, due to the changes in Angular 6, when you subscribe to a `ngForm`'s `valueChanges`, the "plain JS model" values have not yet been updated:

```ts
@Component({
  selector: 'my-form-component',
  template: `
    <ngForm>
      <input type="text" name="First Name" [(ngModel)]="user.firstName" />
    </ngForm>
  `,
})
class MyFormComponent {
  user: User = { firstName: '' };

  @ViewChild(NgForm) ngForm: NgForm;

  ngAfterViewInit() {
    this.form.valueChanges.subscribe(() => {
      console.log(this.user.firstName);
    });
  }
}
```

When you type `C` in the textbox in the above example, the value that gets logged will still be an empty string `''`. `this.user.firstName` isn't updated until after the `valueChanges` emits. This is intended behavior due to other issues.

## Desired behavior
Because of this, there is currently _no way_ to hook into a change in an `ngForm` and have the updated model value.

To be consistent with template-driven forms, it'd be great if there was an `ngFormChange` event on an `ngForm` that would emit _after_ the model is updated:

```html
@Component({
  selector: 'my-form-component',
  template: `
    <ngForm (ngFormChange)="onFormChange()">
      <input type="text" name="First Name" [(ngModel)]="user.firstName" />
    </ngForm>
  `,
})
class MyFormComponent {
  user: User = { firstName: '' };

  onFormChange() {
    console.log(this.user.firstName);
  }
}
```

This is similar to the original code, but the proposal would be that this event is fired after `this.user.firstName` is updated, so now it'd `console.log` `C` when you first type `C`, etc.

I don't think there is a sane `$event` value to pass in in this case... we could send in the form value based on `name` values (cuz we can't get the form model value in the `ngForm`). Could maybe just not send anything. The real issue this is aiming to solve is the _timing_.

## Minimal reproduction of the problem with instructions
Here's an example of `valueChanges` being updated "late" in Angular 6:

https://stackblitz.com/edit/angular-6-value-changes-v2?file=src%2Fapp%2Fapp.component.ts

(Compared to how it worked in Angular 5: https://stackblitz.com/edit/angular-5-value-changes-v2?file=src%2Fapp%2Fapp.component.ts)

## What is the motivation / use case for changing the behavior?
We find it _very_ useful in our code to have a way to know when a template driven form is updated (we've found it 100x easier to use template forms over reactive forms, and we get better type hinting), and we relied on `valueChanges` in Angular 5 that got broke when we updated to Angular 6. I understand why the change would be made, but that doesn't mean an `ngFormChange` type event wouldn't still be very useful.

See: https://github.com/angular/angular/issues/24312

## Environment

<pre><code>
Angular version: 6+

Browser:
- [x] Chrome (desktop) version latest
- [x] Firefox version latest

 
For Tooling issues:
- Node version: 10, 8  
- Platform:  Mac, Windows

Others:

</code></pre>

**Comment by kara** [comment]
@CWSpear Is there any reason you couldn't use the form value that's emitted? It should always be up to date, and is definitely preferred over using the model value (mixing template-driven / reactive techniques).

e.g. https://stackblitz.com/edit/angular-6-value-changes-v2-fgrsrg?file=src/app/app.component.ts

**Comment by CWSpear** [comment]
The idea is not to use a map operation. You lose a lot of the pros of using templates if you start doing something like that.

The idea of an `ngFormChange` output is to avoid having to access stuff in the code (similar to `ngModelChange`). i.e. `(ngFormChange)=onFormChange(model)`.

Is it so strange to everyone that you'd want all of the benefits of a template-driven form, but still be able to know when the model was updated by the form (_after_ the model was updated)?

**Comment by CWSpear** [comment]
Maybe I approached this wrong by opening up the ticket with a specific implementation suggestion. 

Let's take a step back.

The goal here is to be able to use template-driven forms (which me and our team think are much more appropriate and easier to use for what we're doing) and still be able to know when our model has changed (and have the latest changes available in the same cycle). 

We did this in Angular 5 and things were great. But now with the changes, there is no way to know that the model changed _and_ have the most up-to-date model (without trying to map the model to the `name` values).

Since it was functionality that existed, it's not technically difficult to implement (you could make up any sort of new emitter and add it where the `valueChanges` _used_ to emit in Angular 5) and let people choose. That's maybe not ideal, but technically, it's a fairly simple and straightforward solution (just need to emit an event after the view-model is updated).

My specific implementation that I suggested was a way to keep it feeling template-y and consistent.

**Comment by Emeryao** [comment]
+1
I'm facing the same issue as @CWSpear 
Needs this feature please

**Comment by CWSpear** [comment]
@kara I appreciate you triaging and responding, but what would the next steps be here? Is it something that others think has merit? Should I re-open a ticket that has a more general approach to the issue rather than a suggested solution? Would it benefit us if I worked on a PR to show it in action?

**Comment by sar13** [comment]
I have the same issue as well..
I used to rely on the updated model value in  this.form.valueChanges.subscribe ( which used to work .

**Comment by CWSpear** [comment]
@sar13 Yeah, I haven't been able to get a lot of traction going with this, yet. I wanted to test the waters and see if I spent to the time to make a PR, it might actually be merged. Or at least discuss an acceptable approach/compromise.

But it seems like something like this would be immensely useful, I may just try to get a PR up cuz code speaks louder. I know the Angular team is super busy and gets tons and tons of issues... I'm not really sure the best way to go about this sort of thing.

**Comment by sar13** [comment]
Thanks @CWSpear for taking this forward.I am eager to see if we get the behavior back or rewrite a bunch of code to make it working which i don't prefer.

**Comment by pcroc** [comment]
Does this workaround work for you?

```
  ngAfterViewInit() {
    this.form.valueChanges.pipe(debounceTime(0)).subscribe(() => {
      console.log(this.user.firstName);
    });
  }
```

**Comment by CWSpear** [comment]
That's not exactly what we're doing now, but it has the same effect. It makes the updates happen next tick, so now things are happening asynchronously. It works out okay for the user in most cases, but makes it _much_ harder to test and can really start causing problems when you want to do things synchronously based on changes.
